src/main/java/io/cdap/plugin/batch/sink/RecordToBSONWritableTransformer.java

```diff
@@ -0,0 +1,294 @@
+          break;
+        case DECIMAL:
+          BigDecimal decimal = Objects.requireNonNull(record.getDecimal(field.getName()));
+          bsonBuilder.add(field.getName(), new Decimal128(decimal));
+          break;
+        default:
+          throw new UnexpectedFormatException("Field type '" + fieldLogicalType + "' is not supported.");
+      }
+    }
+    return bsonBuilder.get();
+  }
+
+  private ObjectId extractObjectId(Object value) {
+    if (value instanceof byte[]) {
+      return new ObjectId((byte[]) value);
+    } else if (value instanceof ByteBuffer) {
+      byte[] idBytes = ((ByteBuffer) value).array();
+      return new ObjectId(idBytes);
+    } else if (value instanceof String) {
+      return new ObjectId((String) value);
+    }
+
+    throw new UnexpectedFormatException("Invalid ID field value: " + value);
+  }
+
+  private Object extractValue(Object value, Schema fieldSchema) {
+    if (value == null) {
+      // Return 'null' value as it is
+      return null;
+    }
+
+    Schema nonNullableSchema = fieldSchema.isNullable() ? fieldSchema.getNonNullable() : fieldSchema;
+    Schema.Type fieldType = nonNullableSchema.getType();
+    switch (fieldType) {
+      case BOOLEAN:
+        return (Boolean) value;
+      case INT:
+        return (Integer) value;
+      case DOUBLE:
+        return (Double) value;
+      case FLOAT:
+        return (Float) value;
+      case BYTES:
+        if (value instanceof ByteBuffer) {
+          return Bytes.getBytes((ByteBuffer) value);
+        } else {
+          return (byte[]) value;
+        }
+      case LONG:
+        return (Long) value;
+      case STRING:
+        return (String) value;
+      case MAP:
+        return extractMap(value, nonNullableSchema);
+      case ARRAY:
+        return extractArray(value, nonNullableSchema);
+      case RECORD:
+        return extractRecord(value, null);
+      case UNION:
+        return extractUnion(value, nonNullableSchema);
+      case ENUM:
+        String enumValue = (String) value;
+        if (!Objects.requireNonNull(nonNullableSchema.getEnumValues()).contains(enumValue)) {
+          throw new UnexpectedFormatException(String.format("Value '%s' is not enum value. Enum values are: '%s'",
+                                                            enumValue, nonNullableSchema.getEnumValues()));
+        }
+
+        return enumValue;
+      default:
+        throw new UnexpectedFormatException("Field type '" + fieldType + "' is not supported.");
+    }
+  }
+
+  private DBObject extractMap(Object value, Schema schema) {
+    if (value == null) {
+      // Return 'null' value as it is
+      return null;
+    }
+    Map.Entry<Schema, Schema> mapSchema = Objects.requireNonNull(schema.getMapSchema());
+    if (mapSchema.getKey().getType() != Schema.Type.STRING) {
+      throw new UnexpectedFormatException("Key of the 'map' schema must be of the 'string' type.");
+    }
+    if (!(value instanceof Map)) {
```

Comment:
the value from a StructuredRecord should always match what the schema is, otherwise there is no point in the schema. Don't need to check this.

Reply:
Removed.